src/protocol/wire_format.rs

```diff
@@ -57,30 +59,91 @@ uint_wire_format_impl!(u16);
 uint_wire_format_impl!(u32);
 uint_wire_format_impl!(u64);
 
-// The 9P protocol requires that strings are UTF-8 encoded.  The wire format is a u16
-// count |N|, encoded in little endian, followed by |N| bytes of UTF-8 data.
-impl WireFormat for String {
-    fn byte_size(&self) -> u32 {
-        (mem::size_of::<u16>() + self.len()) as u32
-    }
+/// A 9P protocol string.
+///
+/// The string is always valid UTF-8 and 65535 bytes or less (enforced by `P9String::new()`).
+///
+/// It is represented as a C string with a terminating 0 (NUL) character to allow it to be passed
+/// directly to libc functions.
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub struct P9String {
```

Comment:
Should we use `repr(C)` or `repr(transparent)` here?

Reply:
The `struct P9String` itself doesn't need to match any particular layout (like it would if we wanted to use zerocopy or similar crates), since it's only used as a temporary internal representation of the string - it gets serialized/deserialized to/from the bytes that make up the 9p on-wire protocol in the `WireFormat` impl below.

Reply:
LGTM